cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/artifact/ArtifactInspector.java

```diff
@@ -387,6 +407,31 @@ private String getPluginDescription(Class<?> cls) {
     return annotation == null ? "" : annotation.value();
   }
 
+  /**
+   * Returns the metadata mutation for this plugin, return {@code null} if no metadata annotation is there
+   */
+  @Nullable
+  private MetadataMutation getMetadataMutation(PluginId pluginId, Class<?> cls) {
+    Metadata annotation = cls.getAnnotation(Metadata.class);
+    if (annotation == null) {
+      return null;
+    }
+    MetadataScope metadataScope =
+      pluginId.getNamespaceId().equals(NamespaceId.SYSTEM) ? MetadataScope.SYSTEM : MetadataScope.USER;
+    Set<String> tags = new HashSet<>(Arrays.asList(annotation.tags()));
+    MetadataProperty[] metadataProperties = annotation.properties();
+    Map<String, String> properties = new HashMap<>();
+    Arrays.asList(metadataProperties).forEach(property -> properties.put(property.key(), property.value()));
+
+    // if both tags and properties are empty, this means no actual metadata will need to be created
+    if (tags.isEmpty() && properties.isEmpty()) {
+      return null;
+    }
+    return new MetadataMutation.Create(pluginId.toMetadataEntity(),
```

Comment:
what happens if the tags or properties are invalid? At what point will it fail and in what way? Will none of the tags and properties be set or will just the invalid ones be skipped?

Reply:
Just noticed our internal batch call on metadata does not do any validation on the mutation, probably due to these are all system metadata added by us. I added the validation before create this mutation so deploy will fail if the metadata is invalid